### PR TITLE
Ci

### PR DESCRIPTION
**Pull request by JimmyTron**
# CI Workflow

A CI workflow for testing the functionality of the dockerized application and also the endpoints

Issue:
image

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New workflow 
- [ ] This change requires a documentation update

## How Has This Been Tested?

[CI Django & Postgres Tests](https://github.com/SpaceyaTech/blog/actions/workflows/django-postgres-ci.yml)

[Jambo](https://github.com/SpaceyaTech/mastori/actions/workflows/jambo.yaml)

**Test Configuration**:

```sql
          python manage.py migrate
          python manage.py test
```

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .github/workflows/docker ci.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Build the Docker image
+      run: docker build . --file Dockerfile --tag my-image-name:$(date +%s)
+
+    - name: Run the Docker container
+      run: docker run --name container-name -d -p 80:8000 my-image-name
+
```
Comment (JimmyTron) [inline .github/workflows/docker ci.yml:23]: image

**Review by JimmyTron — COMMENTED** [review]
Some refactoring needed here

File: .github/workflows/docker ci.yml (hunk 1 above)
Comment (Gibson-Gichuru) [reply .github/workflows/docker ci.yml:23]: running `docker build . --file Dockerfile --tag my-image-name:$(date +%s)` builds an image with a tagname such as `my-image-name:1679807164` but  there is no way to tell what value `$(date +%s)` will produce. so running the image with `docker run  ........... my-image-name` resolves to `my-image-name:latest` which in this case does not exist. I would advise to just use a simple tag name

File: .github/workflows/docker ci.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    - name: Check if the container is up and running
+      run: docker ps | grep container-name
+
+    - name: Check if the application is running inside the container
+      run: |
+        docker exec container-name netstat -lnp | grep 8000
```
Comment (Gibson-Gichuru) [inline .github/workflows/docker ci.yml:29]: The application image only have steps on how to package the application. running  the container will not start the application. A simpler way would be setting up the environment required by all containers. then let docker compose handle the heavy lifting. All you will need to do is just test if the application is running

Comment (JimmyTron) [reply .github/workflows/docker ci.yml:29]: you could push the changes to this pr.

**Comment by mosesmbadi** [comment]
I have two comments on this, 
1. I would like to advise all the contributors to avoid using placeholder texts/information while submitting a pull request. This will help understand the pr in detail. Please make them as descriptive as possible.
2. I have a few errors on my end and before I move to try and solve them I need to understand if anyone else is facing this error. I have only tested the containerized application where I got these errors: One, the styles and I guess media directory is not loaded hence the styling is missing in the application when ran and two some endpoints are not working ie `/login`, `/mastori ` and  `/account`

Please comment if you are also facing these errors when running the container or if you know a way to solve them. Below is  a log from the container;

(and I just realized our container is initialized with Unix epoch clock!)

1970-01-01 03:00:00 time="2023-03-29T06:57:42+03:00" level=warning msg="Found multiple config files with supported names: /opt/SpaceYaTech/mastori-backend/compose.yml, /opt/SpaceYaTech/mastori-backend/docker-compose.yml"
1970-01-01 03:00:00 time="2023-03-29T06:57:42+03:00" level=warning msg="Using /opt/SpaceYaTech/mastori-backend/compose.yml"
1970-01-01 03:00:00 time="2023-03-29T06:57:42+03:00" level=warning msg="Found multiple config files with supported names: /opt/SpaceYaTech/mastori-backend/compose.yml, /opt/SpaceYaTech/mastori-backend/docker-compose.yml"
1970-01-01 03:00:00 time="2023-03-29T06:57:42+03:00" level=warning msg="Using /opt/SpaceYaTech/mastori-backend/compose.yml"
2023-03-29 06:54:35 mastori-backend-database-1     | The files belonging to this database system will be owned by user "postgres".
2023-03-29 06:54:35 mastori-backend-database-1     | This user must also own the server process.
2023-03-29 06:54:35 mastori-backend-database-1     | 
2023-03-29 06:54:35 mastori-backend-database-1     | The database cluster will be initialized with locale "en_US.utf8".
2023-03-29 06:54:35 mastori-backend-database-1     | The default database encoding has accordingly been set to "UTF8".
2023-03-29 06:54:35 mastori-backend-database-1     | The default text search configuration will be set to "english".
2023-03-29 06:54:35 mastori-backend-database-1     | 
2023-03-29 06:54:35 mastori-backend-database-1     | Data page checksums are disabled.
2023-03-29 06:54:35 mastori-backend-database-1     | 
2023-03-29 06:54:35 mastori-backend-database-1     | fixing permissions on existing directory /var/lib/postgresql/data ... ok
2023-03-29 06:54:35 mastori-backend-database-1     | creating subdirectories ... ok
2023-03-29 06:54:35 mastori-backend-database-1     | selecting dynamic shared memory implementation ... posix
2023-03-29 06:54:35 mastori-backend-database-1     | selecting default max_connections ... 100
2023-03-29 06:54:35 mastori-backend-database-1     | selecting default shared_buffers ... 128MB
2023-03-29 06:54:35 mastori-backend-database-1     | selecting default time zone ... UTC
2023-03-29 06:54:35 mastori-backend-database-1     | creating configuration files ... ok
2023-03-29 06:54:35 mastori-backend-database-1     | running bootstrap script ... ok
2023-03-29 06:54:35 mastori-backend-database-1     | sh: locale: not found
2023-03-29 06:54:35 mastori-backend-database-1     | 2023-03-29 03:54:35.900 UTC [30] WARNING:  no usable system locales were found
2023-03-29 06:54:36 mastori-backend-database-1     | performing post-bootstrap initialization ... ok
2023-03-29 06:54:36 mastori-backend-database-1     | initdb: warning: enabling "trust" authentication for local connections
2023-03-29 06:54:36 mastori-backend-database-1     | initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.
2023-03-29 06:54:36 mastori-backend-database-1     | syncing data to disk ... ok
2023-03-29 06:54:36 mastori-backend-database-1     | 
2023-03-29 06:54:36 mastori-backend-database-1     | 
2023-03-29 06:54:36 mastori-backend-database-1     | Success. You can now start the database server using:
2023-03-29 06:54:36 mastori-backend-database-1     | 
2023-03-29 06:54:36 mastori-backend-database-1     |     pg_ctl -D /var/lib/postgresql/data -l logfile start
2023-03-29 06:54:36 mastori-backend-database-1     | 
2023-03-29 06:54:36 mastori-backend-database-1     | waiting for server to start....2023-03-29 03:54:36.984 UTC [36] LOG:  starting PostgreSQL 15.2 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r4) 12.2.1 20220924, 64-bit
2023-03-29 06:54:36 mastori-backend-database-1     | 2023-03-29 03:54:36.986 UTC [36] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2023-03-29 06:54:36 mastori-backend-database-1     | 2023-03-29 03:54:36.994 UTC [39] LOG:  database system was shut down at 2023-03-29 03:54:36 UTC
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:36.999 UTC [36] LOG:  database system is ready to accept connections
2023-03-29 06:54:37 mastori-backend-database-1     |  done
2023-03-29 06:54:37 mastori-backend-database-1     | server started
2023-03-29 06:54:37 mastori-backend-database-1     | CREATE DATABASE
2023-03-29 06:54:37 mastori-backend-database-1     | 
2023-03-29 06:54:37 mastori-backend-database-1     | 
2023-03-29 06:54:37 mastori-backend-database-1     | /usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*
2023-03-29 06:54:37 mastori-backend-database-1     | 
2023-03-29 06:54:37 mastori-backend-database-1     | waiting for server to shut down....2023-03-29 03:54:37.620 UTC [36] LOG:  received fast shutdown request
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.623 UTC [36] LOG:  aborting any active transactions
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.626 UTC [36] LOG:  background worker "logical replication launcher" (PID 42) exited with exit code 1
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.626 UTC [37] LOG:  shutting down
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.629 UTC [37] LOG:  checkpoint starting: shutdown immediate
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.694 UTC [37] LOG:  checkpoint complete: wrote 918 buffers (5.6%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.016 s, sync=0.040 s, total=0.067 s; sync files=250, longest=0.004 s, average=0.001 s; distance=4222 kB, estimate=4222 kB
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.700 UTC [36] LOG:  database system is shut down
2023-03-29 06:54:37 mastori-backend-database-1     |  done
2023-03-29 06:54:37 mastori-backend-database-1     | server stopped
2023-03-29 06:54:37 mastori-backend-database-1     | 
2023-03-29 06:54:37 mastori-backend-database-1     | PostgreSQL init process complete; ready for start up.
2023-03-29 06:54:37 mastori-backend-database-1     | 
2023-03-29 06:54:47 mastori-backend-application-1  | [2023-03-29 03:54:47 +0000] [1] [INFO] Starting gunicorn 20.1.0
2023-03-29 06:54:47 mastori-backend-application-1  | [2023-03-29 03:54:47 +0000] [1] [INFO] Listening at: http://0.0.0.0:8000 (1)
2023-03-29 06:54:47 mastori-backend-application-1  | [2023-03-29 03:54:47 +0000] [1] [INFO] Using worker: sync
2023-03-29 06:54:47 mastori-backend-application-1  | [2023-03-29 03:54:47 +0000] [7] [INFO] Booting worker with pid: 7
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/css/bootstrap.min.css
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/css/bootstrap-tweaks.css
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/css/default.css
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/css/prettify.css
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/jquery-3.5.1.min.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/ajax-form.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/csrf.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/bootstrap.min.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/prettify-min.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/default.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/jquery-3.5.1.min.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/ajax-form.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/csrf.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/bootstrap.min.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/prettify-min.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /static/rest_framework/js/default.js
2023-03-29 06:56:55 mastori-backend-application-1  | Not Found: /favicon.ico
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.747 UTC [1] LOG:  starting PostgreSQL 15.2 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r4) 12.2.1 20220924, 64-bit
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.748 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.748 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.754 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.762 UTC [52] LOG:  database system was shut down at 2023-03-29 03:54:37 UTC
2023-03-29 06:54:37 mastori-backend-database-1     | 2023-03-29 03:54:37.770 UTC [1] LOG:  database system is ready to accept connections
2023-03-29 06:58:04 mastori-backend-application-1  | Not Found: /mastori
2023-03-29 06:58:24 mastori-backend-application-1  | Not Found: /mastori
2023-03-29 06:58:25 mastori-backend-application-1  | Not Found: /favicon.ico
2023-03-29 06:58:34 mastori-backend-application-1  | Method Not Allowed: /register/
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/css/bootstrap.min.css
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/css/bootstrap-tweaks.css
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/css/prettify.css
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/css/default.css
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/js/ajax-form.js
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/js/jquery-3.5.1.min.js
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/js/csrf.js
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/js/bootstrap.min.js
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/js/prettify-min.js
2023-03-29 06:58:35 mastori-backend-application-1  | Not Found: /static/rest_framework/js/default.js
2023-03-29 06:58:43 mastori-backend-application-1  | Not Found: /login
2023-03-29 06:59:04 mastori-backend-application-1  | Not Found: /account
2023-03-29 06:59:37 mastori-backend-database-1     | 2023-03-29 03:59:37.819 UTC [50] LOG:  checkpoint starting: time
2023-03-29 06:59:41 mastori-backend-database-1     | 2023-03-29 03:59:41.978 UTC [50] LOG:  checkpoint complete: wrote 44 buffers (0.3%); 0 WAL file(s) added, 0 removed, 0 recycled; write=4.128 s, sync=0.014 s, total=4.159 s; sync files=12, longest=0.006 s, average=0.002 s; distance=252 kB, estimate=252 kB
`

**Comment by JimmyTron** [comment]
> I have two comments on this,
> 
> 1. I would like to advise all the contributors to avoid using placeholder texts/information while submitting a pull request. This will help understand the pr in detail. Please make them as descriptive as possible.
> 2. I have a few errors on my end and before I move to try and solve them I need to understand if anyone else is facing this error. I have only tested the containerized application where I got these errors: One, the styles and I guess media directory is not loaded hence the styling is missing in the application when ran and two some endpoints are not working ie `/login`, `/mastori ` and  `/account`
> 
> Please comment if you are also facing these errors when running the container or if you know a way to solve them. Below is a log from the container;
>  

@Gibson-Gichuru On the missing styles we had seen this earlier there was the modules we installed

**Pull request closed without merging**